lib/route/ROUTING.md

```diff
@@ -0,0 +1,446 @@
+page for us to use.
+
+#### Obelisk live development environment
+
+For best results, this tutorial will work best if you work through it using a fresh created
+Obelisk application. To do so, ensure the `ob` command is installed and run the following:
+
+```shell
+$ mkdir ob-routes-tutorial
+$ cd ob-routes-tutorial
+$ ob init
+```
+
+After `ob init` has done its thing, use the live development environment provided by `ob`:
+
+```shell
+$ ob run
+```
+
+### Defining our route as a type
+
+When we define this route, we will need to tell Obelisk what want to do with that number. We
+_could_ leave it as `Text`, but we know it needs to be a number. We could encode to a bare `Int`,
+but we're responsible Haskellers so we will define a `newtype` in `Common.Api` to wrap the number
+for greater type safety and to be more descriptive:
+
+```haskell
+newtype UserId = UserId { unUserId :: Int }
+```
+
+To create the route itself, we add a constructor to the `FrontendRoute` GADT in `Common.Route`.
+
+The purpose of this constructor is to give us something to match on when we're deciding which
+code to run, it will also be used to help build type-safe links within our application. As well
+as declare what the types are that will be pulled from the route.
+
+Add the following constructor to `FrontendRoute` in `Common.Route`:
+
+```haskell
+  FrontendRoute_User :: FrontendRoute UserId
+```
+
+> The naming of the route 'FrontendRoute_User' is an Obelisk convention of including the type
+> name in individual constructor names. This helps disambiguate the code at the 'cost' of a few
+> extra keystrokes. As an example:
+>
+> ```haskell
+> data Foo
+>   = Foo_ConstructorA
+>   | Foo_ConstructorB
+> ```
+
+### Building our route from segments
+
+Next we need to build the `SegmentResult` that will tell Obelisk how this route is to be encoded
+in either direction.
+
+Broadly speaking, the Obelisk router breaks down incoming routes into segments that are given to
+our `Encoder` to find a match. We define these segments using the `SegmentResult` type and add
+them to the case expression in `Common.Route` to be matched against.
+
+Add our new constructor to the following function in `Common.Route` as another pattern in the
+`case` expression for the frontend routes:
+
+```haskell
+fullRouteEncoder
+  :: Encoder (Either Text) Identity (R (FullRoute BackendRoute FrontendRoute)) PageName
+fullRouteEncoder = mkFullRouteEncoder
+  (FullRoute_Backend BackendRoute_Missing :/ ())
+  (\case
+      BackendRoute_Missing -> PathSegment "missing" $ unitEncoder mempty)
+  (\case
+      FrontendRoute_Main -> PathEnd $ unitEncoder mempty
+      -- Add the new constructor here
+  )
+```
+
+Our route has multiple segments so we need to use the `PathSegment` constructor:
+
+```haskell
+  | PathSegment Text (Encoder check parse a PageName)
+```
+
+Matching on the first segment is done by using the literal value "user" as the first argument to
```

Comment:
I know there's a note below, but somehow "Matching" still doesn't feel like it's the whole story here. This is also where you *declare* the path segment.

Reply:
> We declare the first segment by using the literal value "user" as the first argument to the
constructor that we added to the `case` expression:

Changed it up as we don't match at this point, we are _declaring_ the segment here. :+1: